server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java

```diff
@@ -71,24 +69,18 @@ protected boolean resolveIndex(GetRequest request) {
 
     @Override
     protected ShardIterator shards(ClusterState state, InternalRequest request) {
-        ShardIterator shards = clusterService.operationRouting()
+        final ShardIterator iterator = clusterService.operationRouting()
             .getShards(
                 clusterService.state(),
                 request.concreteIndex(),
                 request.request().id(),
                 request.request().routing(),
                 request.request().preference()
             );
-        // If it is stateless, only route isPromotableToPrimary shards. This is a temporary workaround until a more cohesive solution can be
-        // implemented for search shards.
-        if (DiscoveryNode.isStateless(clusterService.getSettings())) {
-            return new PlainShardIterator(
-                shards.shardId(),
-                shards.getShardRoutings().stream().filter(ShardRouting::isPromotableToPrimary).collect(Collectors.toList())
-            );
-        } else {
-            return shards;
+        if (iterator == null) {
+            return null;
         }
```

Comment:
Yeah it is handled in `org.elasticsearch.action.support.single.shard.TransportSingleShardAction.AsyncSingleAction#start`.

Reply:
I think that is different. The abstract class do have subclasses that return null to have the action executed on the local node. But in this case we use `OperationRouting.getShards`, which returns a non-null instance.

I think we should remove these 3 lines (unless you had a test failure and I missed something when looking at this).

Reply:
To me it seems `preferenceActiveShardIterator` which is used by `OperationRouting.getShards` can return null. `useOnlyPromotableShardsForStateless` which also uses its return value, has some null check as well.

Reply:
Would that not also have NPE'd today? We would then execute it "locally", which may not have the shard and thus NPE.

But let us keep it for now. Perhaps add a comment on `preferenceActiveShardIterator` to clarify the case.

Reply:
> Would that not also have NPE'd today? 

I think not because we are not dereferencing it. We just return it and `AsyncSingleAction` takes care of it. In this PR, we use it to filter the shards.